CSV techniques for mitigations, groups and software

emmanvg opened the pull request:
Generates CSV data for all Groups (GXXXX) and Mitigations (MXXXX) objects matched against the ATT&CK ID (TXXXX) via Relationships. All the content is sorted by the Technique ID first, followed by the Group or Mitigation ID.

In this script the Technique ID is the primary column and it is performed to the complete data set but it might contain design elements relevant to mitre-attack/attack-website#135

emmanvg committed:
Add new script that creates CSVs for technique mitigations or groups

emmanvg committed:
add entry in README.md

emmanvg committed:
fix docstring for method `filter_by_type_and_id`

isaisabel commented on scripts/README.md:
```diff
@@ -7,3 +7,4 @@ This folder contains one-off scripts for working with ATT&CK content. These scri
 | [techniques_from_data_source.py](techniques_from_data_source.py) | Fetches the current ATT&CK STIX 2.0 objects from the ATT&CK TAXII server, prints all of the data sources listed in Enterprise ATT&CK, and then lists all the Enterprise techniques containing a given data source. Run `python3 techniques_from_data_source.py -h` for usage instructions. |
 | [techniques_data_sources_vis.py](techniques_data_sources_vis.py) | Generate the csv data used to create the "Techniques Mapped to Data Sources" visualization in the ATT&CK roadmap. Run `python3 techniques_data_sources_vis.py -h` for usage instructions. | 
 | [diff_stix.py](diff_stix.py) | Create markdown and/or ATT&CK Navigator layers reporting on the changes between two versions of the STIX2 bundles representing the ATT&CK content. For default operation, put [enterprise-attack.json](https://github.com/mitre/cti/blob/master/enterprise-attack/enterprise-attack.json), [mobile-attack.json](https://github.com/mitre/cti/blob/master/mobile-attack/mobile- ...
+| [filter_techniques_by_category.py](filter_techniques_by_category.py) | Fetches the current ATT&CK content expressed as STIX2 and creates spreadsheet matching Techniques with Mitigations or Groups. Run `python3 filter_techniques_by_category.py -h` for usage instructions. |
```
I'm not sure `filter_techniques_by_category` is a very descriptive name. Should probably include something about the output, e.g `technique_mappings_to_csv` or something to that effect.

isaisabel replied:
Also, in the description "matching" should probably be "mapping" since most users seem to refer to relationships as "mappings".

emmanvg replied:
👍

isaisabel commented on scripts/filter_techniques_by_category.py:
```diff
@@ -0,0 +1,180 @@
+def arg_parse():
+    """Function to handle script arguments."""
+    parser = argparse.ArgumentParser(description="Fetches the current ATT&CK content expressed as STIX2 and creates spreadsheet matching Techniques with Mitigations or Groups.")
+    parser.add_argument("-c", "--collection", type=str, required=True, choices=["enterprise_attack", "mobile_attack"], help="Which collection to use (Enterprise, Mobile).")
```
I know "collection" is the correct TAXII term, but ATT&CK typically refers to Enterprise and Mobile as domains. I think users are more likely to recognize the term "domain" than "collection."

isaisabel commented on scripts/filter_techniques_by_category.py:
```diff
@@ -0,0 +1,180 @@
+    """Function to handle script arguments."""
+    parser = argparse.ArgumentParser(description="Fetches the current ATT&CK content expressed as STIX2 and creates spreadsheet matching Techniques with Mitigations or Groups.")
+    parser.add_argument("-c", "--collection", type=str, required=True, choices=["enterprise_attack", "mobile_attack"], help="Which collection to use (Enterprise, Mobile).")
+    parser.add_argument("-o", "--operation", type=str, required=True, choices=["groups", "mitigations"], help="Operation to perform on ATT&CK content.")
```
The use of the word "operation" is unclear to me. Were I not already familiar with the script I'd have no idea what this did. Possibly replace with "--mapping-type" and the help with "Which type of object to output the mappings of" or something along those lines.

isaisabel commented on scripts/filter_techniques_by_category.py:
```diff
@@ -0,0 +1,180 @@
+    return sorted(writable_results, key=lambda x: (x["TID"], x["GID"]))
+
+
+def do_mitigations(ds):
+    """Main logic to match techniques to mitigations"""
+    all_attack_patterns = get_all_techniques(ds)
+    writable_results = []
+
+    for attack_pattern in all_attack_patterns:
+        tid = grab_external_id(attack_pattern)
+
+        # Grabs mitigation relationships for identified techniques
+        relationships = filter_for_term_relationships(ds, "mitigates", attack_pattern.id, source=False)
+
+        for relationship in relationships:
+            # Mitigations are defined in STIX as course-of-action objects
+            mitigation = filter_by_type_and_id(ds, "course-of-action", relationship.source_ref)
+
+            if mitigation:
+                mitigation = mitigation[0]
+                mid = grab_external_id(mitigation)
+                writable_results.append(
+                    {
+                        "TID": tid,
+                        "Technique Name": attack_pattern.name,
+                        "MID": mid,
+                        "Mitigation Name": mitigation.name,
+                        "Mitigation Description": escape_chars(mitigation.description),
+                        "Application": escape_chars(relationship.description),
+                    }
+                )
+    return sorted(writable_results, key=lambda x: (x["TID"], x["MID"]))
```
`do_mitigations` and `do_groups` are nearly identical in functionality. Might be worth refactoring into a single function to adhere to good [DRY principles](https://www.wikiwand.com/en/Don%27t_repeat_yourself).

isaisabel reviewed (changes requested):
This is a solid script. I added a few inline comments, mostly just documentation-level stuff. Additional thoughts:
- Would it be difficult to add a "software" mapping (operation) option? It seems like it should be almost identical to the do_groups functionality. This would be especially easy if `do_mitigations` and `do_groups` were refactored into a single function as proposed in the inline review.
- Should target develop, not master. We do staged releases on this repo.
- Addition of this script should also be documented in the changelog. See [the raw develop changelog](https://raw.githubusercontent.com/mitre-attack/attack-scripts/develop/CHANGELOG.md) for a formatting example.

emmanvg committed:
consolidate `do_groups`, `do_mitigations` into a singular method `do_mapping`

add support to for `software` operation, fix issue not allowing mobile-attack operations

emmanvg committed:
docstring changes made from PR feedback, renamed file to `technique_mappings_to_csv.py`

emmanvg committed:
update CHANGELOG.md

emmanvg commented:
@isaisabel, let me know if I missed anything. Hopefully all your suggestions have been incorporated.

emmanvg committed:
Merge branch 'develop' into csv-techniques-for-mitigations-and-groups

emmanvg committed:
update argument names...

isaisabel commented on scripts/technique_mappings_to_csv.py:
```diff
@@ -0,0 +1,168 @@
+    all_attack_patterns = get_all_techniques(ds, source_name)
+    writable_results = []
+
+    for attack_pattern in all_attack_patterns:
+        # Grabs relationships for identified techniques
+        relationships = filter_for_term_relationships(ds, relationship_type, attack_pattern.id)
```
My only recommended change is to have some sort of progress bar. It takes a very long time (up to 4 minutes in my test for software) to parse all the data and users may think that the script crashed while it was working. Wrapping the iterable with [tqdm](https://github.com/tqdm/tqdm), which we've used elsewhere in this repo, would help show users that the script has not crashed.

Something like this is probably sufficient:

```
for attack_pattern in tqdm(all_attack_patterns, desc="parsing data for techniques"):
```
which yields a CLI like this:
```
$ python3 technique_mappings_to_csv.py -m software -d enterprise_attack
parsing data for techniques:   9%|███████▋                                                                          | 25/266 [00:03<00:47,  5.07it/s]

emmanvg committed:
Update technique_mappings_to_csv.py

Add `tqdm` module to provide estimate of procedure completion
